dss_metrics/nvmftarget_collector.py

```diff
@@ -0,0 +1,140 @@
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+
+import re
+import subprocess
+import socket
+import time
+
+import metrics
+import utils
+
+
+class NVMFTargetCollector():
+    def __init__(
+        self,
+        ustat_binary_path,
+        seconds,
+        num_iterations,
+        whitelist_patterns,
+        filter=False
+    ):
+        self.ustat_path = ustat_binary_path
+        self.nvmf_pid, status = utils.check_spdk_running()
+        self.seconds = seconds
+        self.num_iterations = num_iterations
+        self.whitelist_patterns = whitelist_patterns
+        self.filter = filter
+        self.TYPE = 'target'
+
+    def poll_statistics(self, metrics_data_buffer):
+        try:
+            cmd = (
+                self.ustat_path + ' -p '
+                + str(self.nvmf_pid) + ' '
+                + str(self.seconds) + ' '
+                + str(self.num_iterations)
+            )
+            proc = subprocess.Popen(cmd.split(' '), stdout=subprocess.PIPE)
+        except Exception as e:
+            print(f'Caught exception while running USTAT {str(e)}')
+            return
+
+        subsystem_num_to_nqn_map = {}
+        drive_num_to_drive_serial_map = {}
+
+        while True:
+            line = proc.stdout.readline().decode('utf-8')
+            if not line:
+                break
+            line = line.strip()
+            if line:
+                try:
+                    full_key, value = line.split('=')
+                    whitelist_match = False
+
+                    # filter using whitelist patterns if required
+                    if self.filter:
+                        for regex in self.whitelist_patterns:
+                            if re.match(regex, full_key):
+                                whitelist_match = True
+                                break
+
+                    # check if value is a valid promotheus metric value
+                    valid_value_flag = value.replace('.', '').isdigit()
+
+                    fields = full_key.split('.')
+                    subsystem_num = fields[1]
+                    component = fields[2]
+                    metric_name = fields[-1]
+
+                    if component == 'id' and 'nqn' in metric_name:
+                        subsystem_num_to_nqn_map[subsystem_num] = value
+                    elif 'drive' in component and 'serial' in metric_name:
+                        drive_num_to_drive_serial_map[component] = value
+                    elif 'ip' in metric_name:
+                        valid_value_flag = False
+
+                    tags = {}
+                    tags['cluster_id'] = 'c01'
```

Comment:
Define this value as a global variable or the config file. We may need to find a mechanism on how to fill the cluster id when more clusters are added